### PR TITLE
New wizard

### PR DESCRIPTION
**Pull request by fmoc**
This PR presents a complete reimplementation of the existing wizard. Some preparation has been merged already, e.g., #9426 or #9422. The current state works and sets up accounts with servers that use OAuth2 and classic HTTP Basic authentication.

Some functionality has not been implemented yet:

- Some initial work on the "advanced sync options" was removed again, as it would have to be rewritten again for the new spaces support. Proper theme-based styling will be implemented in a future PR.
- Styling has not been implemented yet. The design is inspired by #5030 already, though. The fine work should be done in a subsequent PR.
- Some features of the old wizard have not been ported yet, e.g., options whether to enable VFS (see the related TODOs).

The PR's commit history is still a mess at this point and will be cleaned up soon. Also, some copyright headers are still missing, etc. The code is ready for review, though.

### CODE REVIEW

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.cpp:23]: Can we use the .ui?

Mixing manual layouting and a .ui usually doesn't end well

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.cpp:52]: Is this needed, the default state should be off, also as far as I know all logic here is in the painter so once its hidden there are no cycles to save.

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.cpp:111]: ???

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.cpp:117]: ???

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.cpp:129]: static_cast?

File: src/gui/newwizard/SetupWizardWindow.h
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardWindow.h:67]: Please use parenting.
There is no reason to use shared pointers.
Maybe scoped pointers but with Qt there is usually no need for smart pointers, you'll only end up with undefined lifetimes.

File: src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp
Comment (TheOneRing) [inline src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp:28]: .toInt?

File: src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp
Comment (TheOneRing) [inline src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp:33]: Yes I think the line is unreachable.

File: src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp
Comment (TheOneRing) [inline src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp:45]: You should display the refactored redirect url? I assume?

File: src/gui/newwizard/pages/AbstractSetupWizardPage.h
Comment (TheOneRing) [inline src/gui/newwizard/pages/AbstractSetupWizardPage.h:14]: Why a slot when it is directly called?
Maybe rather define a signal that can be subscribed to instead of requiring to define an empty virtual function?

File: src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp
Comment (TheOneRing) [inline src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp:19]: Shouldn't this be handled by the event filter in the window?

File: src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp
Comment (TheOneRing) [inline src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp:24]: fromUserImput

File: src/gui/owncloudgui.cpp
Comment (TheOneRing) [inline src/gui/owncloudgui.cpp:56]: The folder is not automatically created as we want to be able to detect a removed folder.

File: src/gui/owncloudgui.cpp
Comment (TheOneRing) [inline src/gui/owncloudgui.cpp:53]: Move this to the FolderManager?

File: src/gui/owncloudgui.cpp
Comment (TheOneRing) [inline src/gui/owncloudgui.cpp:1046]: We should investigate why this is needed.

File: src/gui/owncloudgui.cpp
Comment (TheOneRing) [inline src/gui/owncloudgui.cpp:1053]: Should happen in any case?

File: src/gui/newwizard/SetupWizardController.h
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardController.h:16]: Any gain of not inheriting SetupWizardWindow?
All I can see is the quite unusual showUi function?

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/gui/newwizard/SetupWizardAccountBuilder.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardAccountBuilder.cpp:43]: I was more interested in making sure the builder is in a proper state, but I guess checking for validity is a good idea.

I'm not sure whether to use asserts at all, honestly. I was thinking about making all setters return `bool`, replacing them with `OC_ENSURE` and handling errors in some way. What do you think about that?

File: src/gui/newwizard/SetupWizardController.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardController.cpp:40]: I think it's useful for testing, because one can see the transition "page". I'd clean that up before merging the PR.

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardWindow.cpp:52]: I wasn't sure whether to use it or not. I thought, since it's offered, why not use it, it definitely can't hurt... I'll check that again.

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardWindow.cpp:111]: The first value is something I picked from Qt creator, some integer maximum of some kind, I'd say.

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardWindow.cpp:129]: Makes `clang-tidy` unhappy, because we're casting down from a base class to a derived one.

File: src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp
Comment (fmoc) [reply src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp:19]: In the current state the signal doesn't seem to be propagated to the window. But I've changed a lot of things since. I'll try.

File: src/gui/owncloudgui.cpp
Comment (fmoc) [reply src/gui/owncloudgui.cpp:56]: So you're saying this line of code should be there? I couldn't find how the old wizard solved this in a reasonable amount of time, and just hacked some `mkdir` into a location that seemed sensible.

File: src/gui/owncloudgui.cpp
Comment (fmoc) [reply src/gui/owncloudgui.cpp:1046]: Should I leave a TODO? Should we open an issue?

File: src/gui/owncloudgui.cpp
Comment (fmoc) [reply src/gui/owncloudgui.cpp:1069]: Can do that. My best guess is that it was done like this before was that the wizard is modal to the settings, and they wanted to show both. I'm curious to find out how the windows will behave when calling `raiseDialog` on the modal dialog.

File: src/gui/newwizard/SetupWizardController.h
Comment (fmoc) [reply src/gui/newwizard/SetupWizardController.h:16]: Separation of concerns. The controller does not do any UI-specific code, just business logic. The window and the pages implement UI-specific details. The builder serves as a data storage facility. Inheriting would violate SoC. I'd rather not mix both controller and frontend code.

The chosen architecture is similar to MVC.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/10974/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/10974/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/10974/guiReportUpload/serverlog.log)

File: src/gui/owncloudgui.cpp
Comment (TheOneRing) [reply src/gui/owncloudgui.cpp:1046]: Please open an issue.

File: src/gui/newwizard/SetupWizardController.h
Comment (TheOneRing) [reply src/gui/newwizard/SetupWizardController.h:16]: Then maybe make the widget accessible instead of implementing the showUi()

File: src/gui/newwizard/SetupWizardWindow.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardWindow.cpp:136]: Simulating the click works just fine. I can move the callback bound to the button of the signal to become a slot of the class, if you prefer that.

File: src/gui/newwizard/SetupWizardWindow.h
Comment (fmoc) [reply src/gui/newwizard/SetupWizardWindow.h:67]: In this use case, the shared pointer is just used to auto-delete the auto-generated class in the destructor. Since we don't intend to ever implement move behavior, a scoped pointer would work, too. However, scoped pointers require the class declaration to be included, which means including the auto-generated headers in those headers:

```
error: invalid application of ‘sizeof’ to incomplete type ‘Ui::AccountConfiguredWizardPage’
```

I don't think you can use parenting properly with those auto-generated classes, can you? They are not `QObject`s, but plain classes.

File: src/gui/newwizard/pages/AbstractSetupWizardPage.h
Comment (fmoc) [reply src/gui/newwizard/pages/AbstractSetupWizardPage.h:14]: I first had a default implementation in place, but that was only useful for a single class. All other classes need to perform custom setup code.

I think this solution is more reliable, because the child class does not need to know about the parent class.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/10982/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/10982/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/10982/guiReportUpload/serverlog.log)

File: src/gui/owncloudgui.cpp
Comment (fmoc) [reply src/gui/owncloudgui.cpp]: Now that I don't raise the settings dialog but the wizard window, it won't be centered above the settings any more on multi-screen systems. I think that's been the motivation to implement this in the first place.

File: src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp
Comment (fmoc) [reply src/gui/newwizard/pages/ServerUrlSetupWizardPage.cpp:19]: Found it. Have to install the window as its own event filter. Seems a bit awkward at first, but actually makes sense.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/10999/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/10999/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/10999/guiReportUpload/serverlog.log)

File: src/gui/newwizard/Pagination.cpp
Comment (TheOneRing) [reply src/gui/newwizard/Pagination.cpp:77]: int m_currentPage;

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11010/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11010/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11010/guiReportUpload/serverlog.log)

File: src/gui/newwizard/Pagination.cpp
Comment (TheOneRing) [inline src/gui/newwizard/Pagination.cpp:30]: _ui->layout->addWidget

File: src/gui/newwizard/Pagination.cpp
Comment (TheOneRing) [inline src/gui/newwizard/Pagination.cpp:33]: Just skip the parameter

File: src/gui/newwizard/SetupWizardController.cpp
Comment (TheOneRing) [inline src/gui/newwizard/SetupWizardController.cpp:137]: Could be put in to a helper together with a Q_ASSERT(pagePtr)

File: src/gui/newwizard/pages/AbstractSetupWizardPage.h
Comment (TheOneRing) [reply src/gui/newwizard/pages/AbstractSetupWizardPage.h:14]: Then unslotify it, its not used as a slot.
A signal would still be more _Qt_

File: src/gui/accountsettings.cpp
Comment (fmoc) [reply src/gui/accountsettings.cpp:545]: So I'd just remove the entire code that won't work any more (`OwncloudWizard` doesn't exist any more, hence it won't compile as-is) and replace it with a comment?

File: src/gui/newwizard/SetupWizardController.cpp
Comment (fmoc) [reply src/gui/newwizard/SetupWizardController.cpp:137]: The cast is not needed anyway, since the controller doesn't need to parse data off the final page.

File: src/gui/newwizard/Pagination.cpp
Comment (fmoc) [reply src/gui/newwizard/Pagination.cpp:30]: See https://github.com/owncloud/client/pull/9482#discussion_r829403489.

File: src/gui/application.h
Comment (fmoc) [reply src/gui/application.h:69]: I'm not entirely sure, I don't mind having `addNewAccount` in the public interface. I'll remove it for now.

File: src/gui/application.cpp
Comment (fmoc) [reply src/gui/application.cpp:501]: I've checked how this is stored at the moment. We'd need to store whether to autostart or not in our config, not just autodetect it from the system. Then, we can also specify a default value on first run more easily (right now, we can't really tell whether this is a first run or not).

I'd do any modifications in a separate follow-up PR, as it will take some testing to make sure we get it right and it's a slightly complex change.

File: src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp
Comment (TheOneRing) [inline src/gui/newwizard/jobs/ResolveUrlJobFactory.cpp:52]: The error message doesn't make sense when the user rejected the redirect.

File: src/gui/ocwizard_deprecated.h
Comment (TheOneRing) [inline src/gui/ocwizard_deprecated.h:13]: Please also add runtime warnings in order to minimize the chance that we ship it in this state.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11142/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11142/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11142/guiReportUpload/serverlog.log)

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11147/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11147/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11147/guiReportUpload/serverlog.log)

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11148/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11148/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11148/guiReportUpload/serverlog.log)

**Comment by fmoc** [comment]
Just the GUI tests fail at the moment. Shouldn't block the merge.

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11149/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11149/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11149/guiReportUpload/serverlog.log)

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/11172/5/1
GUI Logs: (https://cache.owncloud.com/public/owncloud/client/11172/guiReportUpload/index.html)
Server Logs: (https://cache.owncloud.com/public/owncloud/client/11172/guiReportUpload/serverlog.log)

**Pull request merged**
